### PR TITLE
Add support for wide-column point lookups

### PR DESCRIPTION
**Pull request by ltamasi**
Summary:
The patch adds a new API `GetEntity` that can be used to perform
wide-column point lookups. It also extends the `Get` code path and
the `MemTable` / `MemTableList` and `Version` / `GetContext` logic
accordingly so that wide-column entities can be served from both
memtables and SSTs. If the result of a lookup is a wide-column entity
(`kTypeWideColumnEntity`), it is passed to the application in deserialized
form; if it is a plain old key-value (`kTypeValue`), it is presented as a
wide-column entity with a single default (anonymous) column.
(In contrast, regular `Get` returns plain old key-values as-is, and
returns the value of the default column for wide-column entities, see
#10483 .)

The result of `GetEntity` is a self-contained `PinnableWideColumns` object.
`PinnableWideColumns` contains a `PinnableSlice`, which either stores the
underlying data in its own buffer or holds on to a cache handle. It also contains
a `WideColumns` instance, which indexes the contents of the `PinnableSlice`,
so applications can access the values of columns efficiently.

There are several pieces of functionality which are currently not supported
for wide-column entities: there is currently no `MultiGetEntity` or wide-column
iterator; also, `Merge` and `GetMergeOperands` are not supported, and there
is no `GetEntity` implementation for read-only and secondary instances.
We plan to implement these in future PRs.

Test Plan:
`make check`

### CODE REVIEW

File: db/db_impl/db_impl.cc [hunk 1]
```diff
@@ -1822,6 +1822,28 @@ Status DBImpl::Get(const ReadOptions& read_options,
   return s;
 }
 
+Status DBImpl::GetEntity(const ReadOptions& read_options,
+                         ColumnFamilyHandle* column_family, const Slice& key,
+                         PinnableWideColumns* columns) {
+  if (!column_family) {
+    return Status::InvalidArgument(
+        "Cannot call this method without a column family handle");
```
Comment (akankshamahajan15) [inline db/db_impl/db_impl.cc:1830]: Can we add method name in the `Status` message?

File: db/db_impl/db_impl.cc [hunk 2]
```diff
@@ -1980,31 +2003,46 @@ Status DBImpl::GetImpl(const ReadOptions& read_options, const Slice& key,
...
     // Get value associated with key
     if (get_impl_options.get_value) {
-      if (sv->mem->Get(lkey, get_impl_options.value->GetSelf(), timestamp, &s,
-                       &merge_context, &max_covering_tombstone_seq,
-                       read_options, false /* immutable_memtable */,
-                       get_impl_options.callback,
-                       get_impl_options.is_blob_index)) {
+      if (sv->mem->Get(
+              lkey,
+              get_impl_options.value ? get_impl_options.value->GetSelf()
```
Comment (akankshamahajan15) [inline db/db_impl/db_impl.cc:2008]: For my understanding, if `get_impl_options.get_value` is true can `get_impl_options.value` be `nullptr`?

Comment (ltamasi) [reply db/db_impl/db_impl.cc:2008]: Yes, that's the case with the new `GetEntity` API. So for `Get`, we have `get_value == true && value != nullptr && columns == nullptr`; for `GetEntity`, we have `get_value == true && value == nullptr && columns != nullptr`.

**Review by akankshamahajan15 — APPROVED** [review]
LGTM!

**Pull request closed without merging**